src/storage.ts

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2021 Google Inc. All Rights Reserved.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+import { openDB, DBSchema, IDBPDatabase } from 'idb';
+
+const mainDBName = 'appDB';
+const mainImageName = 'mainImage';
+
+export type FileUniqueID = string;  // generated by the storage layer
+export type Timestamp = number;
+
+export type FileMetadata = {
+  filename: string;
+  mimeType: string;
+  title?: string;
+}
+
+export type FileRecord = {
+  id: FileUniqueID;
+  metadata: FileMetadata;
+  added: Timestamp;
+  blobModified: Timestamp;
+  metadataModified: Timestamp;
+};
+
+interface LlamaDB extends DBSchema {
+  blob: {
+    key: string;
+    value: Blob;
+  };
+  metadata: {
+    key: string;
+    value: FileRecord;
+  };
+}
+
+export default class LlamaStorage {
+  private db: IDBPDatabase<LlamaDB>;
+
+  private constructor(db: IDBPDatabase<LlamaDB>) {
+    console.log('Database initialized');
+    this.db = db;
+  }
+
+  static async create() {
+    return new LlamaStorage(await openDB(mainDBName, undefined /* version */, {
+      upgrade(db, oldVersion, newVersion, transaction) {
+        console.log(`DB update request`);
+        db.createObjectStore('blob');
+        db.createObjectStore('metadata');
+      },
+      // blocked()
+      // blocking()
+      // terminated()
+    }));
+  }
+
+  async add(file: Blob, metadata: FileMetadata): Promise<FileRecord> {
+    const id: FileUniqueID = mainImageName /* TODO: generate unique IDs */;
+    const now = Date.now();
+    const record: FileRecord = {
+      id: id,
+      metadata: metadata,
+      added: now,
+      blobModified: now,
+      metadataModified: now,
+    };
+
+    // TODO: it would be nice if the caller could access the FileRecord
+    // immediately, but then still indicate when the data finished
+    // being stored.
+    await Promise.allSettled([
+      this.db.put('blob', file, id),
+      this.db.put('metadata', record, id),
+    ]);
+    console.log(`successfully added '${id}'`);
+    return record;
+  }
+
+  async list(): Promise<FileRecord[]> {
+    // TODO: https://github.com/jakearchibald/idb#async-iterators
+    throw 'list() is unimplemented';
+  }
+
+  async getFile(id: FileUniqueID): Promise<Blob> {
+    const blob = await this.db.get('blob', id);
+    if (!blob) throw `no blob found for id '${id}'`;
```

Comment:
I'm up for either, but in JavaScript it's fairly idiomatic for storage classes to return `undefined` when no value is associated with a particular key. We're throwing an exception here: is that strictness deliberate?

Reply:
Didn't realize that was idiomatic. The reason I went with this is for cleaner code. It's annoying to have to return `Promise<Blob?>` rather than `Promise<Blob>`, and it means that all the callers have to do null/undefined checks before using the result. Maybe a try/catch isn't much better, but I thought throwing on error was a common thing in promises.

Reply:
Rejecting a promise should be exceptional, i.e. anything you'd throw an exception for. For built-in storage types including arrays, objects, Map/Set etc, getting an entry that doesn't exist just returns `undefined`. It looks like this has also been adopted by some async libraries:

https://github.com/jakearchibald/idb-keyval

The thing that bothers me a bit is that you wouldn't write `Promise<Blob?>`, but rather `Promise<Blob | undefined>`, which is far more verbose. I've been tempted to include an `Optional<>` or `Maybe<>` type in some of my personal projects, which would allow for something like `Promise<Optional<Blob>>`, but wouldn't result in any changes on the calling side.

Reply:
Ok, switched to returning possibly-undefined gross type. I kept the exceptions in `update()`, since setting the return value to undefined in that case (if we even decide to keep that return value around) is much less likely to be checked, and the write error could easily go unnoticed.

(perhaps at some point we can change it to gracefully fall back to doing an "add" instead of an "update" if possible, but I think this is good enough for now)